django/db/models/query.py

```diff
@@ -1486,33 +1485,27 @@ def filter(self, *args, **kwargs):
         set.
         """
         self._not_support_combined_queries("filter")
-        return self._filter_or_exclude(False, args, kwargs)
+        return self._filter_q(Q(*args, **kwargs))
 
     def exclude(self, *args, **kwargs):
         """
         Return a new QuerySet instance with NOT (args) ANDed to the existing
         set.
         """
         self._not_support_combined_queries("exclude")
-        return self._filter_or_exclude(True, args, kwargs)
+        return self._filter_q(~Q(*args, **kwargs))
 
-    def _filter_or_exclude(self, negate, args, kwargs):
-        if (args or kwargs) and self.query.is_sliced:
+    def _filter_q(self, q):
+        if q and self.query.is_sliced:
             raise TypeError("Cannot filter a query once a slice has been taken.")
         clone = self._chain()
         if self._defer_next_filter:
             self._defer_next_filter = False
-            clone._deferred_filter = negate, args, kwargs
+            clone._deferred_filter = q
```

Comment:
There might be some logic relying on this attribute being `tuple[bool, tuple, dict]` and not `Q`.